### PR TITLE
[SPARK-27872][K8S][2.4] Fix executor service account inconsistency

### PR DESCRIPTION
**Pull request by nssalian**
### What changes were proposed in this pull request? 

Similar patch to #24748 but applied to the branch-2.4.
Backporting the fix to releases 2.4.x.

### CODE REVIEW

**Comment by nssalian** [comment]
> Can we have java doc here?

There's no javadoc in the original PR (https://github.com/apache/spark/pull/24748). This is simply a backport. Any improvements can follow in a different PR.

**Comment by nssalian** [comment]
@dongjoon-hyun @erikerlandson please let me know if you have additional comments that I can address.

**Comment by nssalian** [comment]
ping:  @dongjoon-hyun @erikerlandson, please review when you get a chance.

**Comment by nssalian** [comment]
Checking again for this PR:  @dongjoon-hyun @erikerlandson.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -61,10 +61,9 @@ private[spark] object Config extends Logging {
       .stringConf
       .createOptional
 
-  val KUBERNETES_AUTH_DRIVER_CONF_PREFIX =
-      "spark.kubernetes.authenticate.driver"
-  val KUBERNETES_AUTH_DRIVER_MOUNTED_CONF_PREFIX =
-      "spark.kubernetes.authenticate.driver.mounted"
+  val KUBERNETES_AUTH_DRIVER_CONF_PREFIX = "spark.kubernetes.authenticate.driver"
+  val KUBERNETES_AUTH_EXECUTOR_CONF_PREFIX = "spark.kubernetes.authenticate.executor"
+  val KUBERNETES_AUTH_DRIVER_MOUNTED_CONF_PREFIX = "spark.kubernetes.authenticate.driver.mounted"
```
Comment (erikerlandson) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:66]: this may blow the 100-char limit, if I counted right

Comment (nssalian) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:66]: The linters didn't fail so I think we are ok.

**Comment by erikerlandson** [comment]
ok to test

**Comment by nssalian** [comment]
> ok to test

We might need an admin to initiate tests again. ping @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by nssalian** [comment]
Thanks @dongjoon-hyun.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 2]
```diff
@@ -16,6 +16,10 @@
  */
 package org.apache.spark.deploy.k8s
 
+import io.fabric8.kubernetes.api.model.{Container, ContainerBuilder,
+  ContainerStateRunning, ContainerStateTerminated,
+  ContainerStateWaiting, ContainerStatus, Pod, PodBuilder}
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:21]: Please use a single line like the original branch, @nssalian .

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 3]
```diff
@@ -60,4 +64,15 @@ private[spark] object KubernetesUtils {
   }
 
   def parseMasterUrl(url: String): String = url.substring("k8s://".length)
+
+  def buildPodWithServiceAccount(serviceAccount: Option[String], pod: SparkPod): Option[Pod] = {
+      serviceAccount.map { account =>
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:69]: We are using `2-space` indentation.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 4]
```diff
@@ -60,4 +64,15 @@ private[spark] object KubernetesUtils {
...
+        new PodBuilder(pod.pod)
+          .editOrNewSpec()
+          .withServiceAccount(account)
+          .withServiceAccountName(account)
+          .endSpec()
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:74]: Please keep the original formatting.
```scala
.editOrNewSpec()
  .withServiceAccount(account)
  .withServiceAccountName(account)
.endSpec()
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverKubernetesCredentialsFeatureStep.scala [hunk 5]
```diff
@@ -70,15 +71,8 @@ private[spark] class DriverKubernetesCredentialsFeatureStep(kubernetesConf: Kube
...
-        pod = driverServiceAccount.map { account =>
-          new PodBuilder(pod.pod)
-            .editOrNewSpec()
-              .withServiceAccount(account)
-              .withServiceAccountName(account)
-              .endSpec()
-            .build()
-        }.getOrElse(pod.pod))
+      pod.copy(pod = buildPodWithServiceAccount(driverServiceAccount, pod)
+        .getOrElse(pod.pod))
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverKubernetesCredentialsFeatureStep.scala:75]: Why do we need to split this line? The original code is one-liner in `master` branch.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+package org.apache.spark.deploy.k8s.features
+
+import org.apache.spark.deploy.k8s.{KubernetesConf, SparkPod}
+import org.apache.spark.deploy.k8s.Config.KUBERNETES_SERVICE_ACCOUNT_NAME
+import org.apache.spark.deploy.k8s.KubernetesUtils.buildPodWithServiceAccount
+
+private[spark] class ExecutorKubernetesCredentialsFeatureStep(kubernetesConf: KubernetesConf[_])
+  extends KubernetesFeatureConfigStep {
+  private lazy val driverServiceAccount =
+    kubernetesConf.get(KUBERNETES_SERVICE_ACCOUNT_NAME)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala:26]: ditto. Please keep one-liner (the same format with the original code of `master` branch).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    pod.copy(
+      // if not setup by the pod template fallback to the driver's sa,
+      // last option is the default sa.
+      pod = if (Option(pod.pod.getSpec.getServiceAccount).isEmpty) {
+        buildPodWithServiceAccount(driverServiceAccount, pod).getOrElse(pod.pod)
+      } else {
+        pod.pod
+      })
+  }
+  override def getAdditionalPodSystemProperties(): Map[String, String] = Map.empty
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala:38]: Thank you for adding this. BTW, could you add one empty line before line 38?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+  override def getAdditionalKubernetesResources(): Seq[io.fabric8.kubernetes.api.model.HasMetadata]
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala:40]: Please import `io.fabric8.kubernetes.api.model.HasMetadata`.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  = Seq.empty
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/ExecutorKubernetesCredentialsFeatureStep.scala:41]: Then, you don't need break a line like this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala [hunk 10]
```diff
@@ -18,12 +18,17 @@ package org.apache.spark.scheduler.cluster.k8s
 
 import org.apache.spark.deploy.k8s.{KubernetesConf, KubernetesExecutorSpecificConf, KubernetesRoleSpecificConf, SparkPod}
 import org.apache.spark.deploy.k8s.features._
-import org.apache.spark.deploy.k8s.features.{BasicExecutorFeatureStep, EnvSecretsFeatureStep, LocalDirsFeatureStep, MountSecretsFeatureStep}
+import org.apache.spark.deploy.k8s.features.{BasicExecutorFeatureStep, EnvSecretsFeatureStep,
+  ExecutorKubernetesCredentialsFeatureStep,
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala:22]: Please don't break `import` statement. `import` is excluded at line-length style check.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala [hunk 11]
```diff
@@ -18,12 +18,17 @@ package org.apache.spark.scheduler.cluster.k8s
...
+LocalDirsFeatureStep, MountSecretsFeatureStep}
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala:23]: ditto.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala [hunk 12]
```diff
@@ -18,12 +18,17 @@ package org.apache.spark.scheduler.cluster.k8s
...
 
 private[spark] class KubernetesExecutorBuilder(
     provideBasicStep: (KubernetesConf [KubernetesExecutorSpecificConf])
       => BasicExecutorFeatureStep =
       new BasicExecutorFeatureStep(_),
+    provideCredentialsStep: (KubernetesConf [KubernetesExecutorSpecificConf])
+      => ExecutorKubernetesCredentialsFeatureStep =
+    new ExecutorKubernetesCredentialsFeatureStep(_),
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala:31]: indentation?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala [hunk 13]
```diff
@@ -50,8 +55,10 @@ private[spark] class KubernetesExecutorBuilder(
     val volumesFeature = if (kubernetesConf.roleVolumes.nonEmpty) {
       Seq(provideVolumesStep(kubernetesConf))
     } else Nil
+    val credentialsFeatures = Seq(provideCredentialsStep(kubernetesConf))
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesExecutorBuilder.scala:58]: `credentialsFeatures` -> `credentialsFeature`.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/BasicTestsSuite.scala [hunk 14]
```diff
@@ -84,6 +84,13 @@ private[spark] trait BasicTestsSuite { k8sSuite: KubernetesSuite =>
       })
   }
 
+  test("All pods have the same service account by default", k8sTestTag) {
+      runSparkPiAndVerifyCompletion(
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/BasicTestsSuite.scala:88]: indentation?

**Comment by nssalian** [comment]
Thanks @dongjoon-hyun. I shall address the comments in upcoming commits.

**Comment by nssalian** [comment]
@dongjoon-hyun please let me know if you have additional comments. Tests seem to have passed.

**Comment by dongjoon-hyun** [comment]
Got it. Thank you for updates.

**Comment by nssalian** [comment]
@dongjoon-hyun  could we merge this if there aren't additional comments or feedback?

**Comment by erikerlandson** [comment]
I'm OK to merge if @dongjoon-hyun  is

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Sorry for the delay, @nssalian and @erikerlandson .
Merged to branch-2.4.

**Pull request closed without merging**